src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -944,6 +947,15 @@ public function setAttribute($key, $value)
             return $this->setAttributeMarkedMutatedAttributeValue($key, $value);
         }
 
+        if ($value instanceof Expression) {
+            if ($value instanceof CalculableExpression) {
+                $value->setExpectedResult($this->calculateResult($value));
```

Comment:
At property assignment (`$user->counter = DB::calculate(...)`), the expression is set as the property and if it is a Calculable Expression, the result is calculated and added to the `CalculableExpression` object so we can later assign it. The actual assignment happens later on in the routine after the query has completed (just before the `updated`/`saved`/`created` hooks).